src/element.js

```diff
@@ -98,46 +98,91 @@ function addNormalRectPath(ctx, rect) {
   ctx.rect(rect.x, rect.y, rect.w, rect.h);
 }
 
-export function shouldDrawCaption(rect, options) {
+export function shouldDrawCaption(displayMode, rect, options) {
   if (!options || options.display === false) {
     return false;
   }
+  if (displayMode === 'headerBoxes') {
+    return true;
+  }
   const {w, h} = rect;
   const font = toFont(options.font);
   const min = font.lineHeight;
   const padding = limit(valueOrDefault(options.padding, 3) * 2, 0, Math.min(w, h));
   return (w - padding) > min && (h - padding) > min;
 }
 
+export function getCaptionHeight(displayMode, rect, font, padding) {
+  if (displayMode !== 'headerBoxes') {
+    return font.lineHeight + padding * 2;
+  }
+  const captionHeight = font.lineHeight + padding * 2;
+  return rect.h < 2 * captionHeight ? rect.h / 3 : captionHeight;
+}
+
 function drawText(ctx, rect, options, item, levels) {
-  const {captions, labels} = options;
+  const {captions, labels, displayMode} = options;
   ctx.save();
   ctx.beginPath();
   ctx.rect(rect.x, rect.y, rect.w, rect.h);
   ctx.clip();
   const isLeaf = item && (!defined(item.l) || item.l === levels);
   if (isLeaf && labels.display) {
     drawLabel(ctx, rect, options);
-  } else if (!isLeaf && shouldDrawCaption(rect, captions)) {
+  } else if (!isLeaf && shouldDrawCaption(displayMode, rect, captions)) {
     drawCaption(ctx, rect, options, item);
   }
   ctx.restore();
 }
 
 function drawCaption(ctx, rect, options, item) {
-  const {captions, spacing, rtl} = options;
+  const {captions, spacing, rtl, displayMode} = options;
   const {color, hoverColor, font, hoverFont, padding, align, formatter} = captions;
   const oColor = (rect.active ? hoverColor : color) || color;
   const oAlign = align || (rtl ? 'right' : 'left');
   const optFont = (rect.active ? hoverFont : font) || font;
   const oFont = toFont(optFont);
+  const fonts = [oFont];
+  if (oFont.lineHeight > rect.h) {
+    return;
+  }
+  let text = formatter || item.g;
+  const captionSize = measureLabelSize(ctx, [formatter], fonts);
+  if (captionSize.width + 2 * padding > rect.w) {
+    text = sliceTextToFitWidth(ctx, text, rect.w - 2 * padding, fonts);
+  }
+
   const lh = oFont.lineHeight / 2;
   const x = calculateX(rect, oAlign, padding);
   ctx.fillStyle = oColor;
   ctx.font = oFont.string;
   ctx.textAlign = oAlign;
   ctx.textBaseline = 'middle';
-  ctx.fillText(formatter || item.g, x, rect.y + padding + spacing + lh);
+  const y = displayMode === 'headerBoxes' ? rect.y + rect.h / 2 : rect.y + padding + spacing + lh;
+  ctx.fillText(text, x, y);
+}
+
+function sliceTextToFitWidth(ctx, text, width, fonts) {
```

Comment:
Looks like you had another improvement hidden here, maybe some tests for this?

Reply:
Yep my bad, I forgot the test